tidy3d/components/data/monitor_data.py

```diff
@@ -133,7 +133,136 @@ def symmetry_expanded_copy(self) -> AbstractFieldData:
+        coordinates."""
+        tan_dims = self._tangential_dims
+        tan_fields = self._tangential_fields
+        plane_bounds0 = tan_fields["E" + tan_dims[1]].coords[tan_dims[0]].values
+        plane_bounds1 = tan_fields["E" + tan_dims[0]].coords[tan_dims[1]].values
+        return plane_bounds0, plane_bounds1
+
+    @cached_property
+    def _diff_area(self) -> xr.DataArray:
+        """For a 2D monitor data, return the area of each cell in the plane, for use in numerical
+        integrations."""
+        bounds = self._plane_grid_boundaries
+        sizes = [bs[1:] - bs[:-1] for bs in bounds]
+        return xr.DataArray(np.outer(sizes[0], sizes[1]), dims=self._tangential_dims)
+
+    @property
+    def _centered_tangential_fields(self) -> Dict[str, DataArray]:
+        """For a 2D monitor data, get the tangential E and H fields colocated to the cell centers in
+        the 2D plane grid."""
+
+        # Tangential directions and fields
+        tan_dims = self._tangential_dims
+        tan_fields = self._tangential_fields
+
+        # Plane center coordinates
+        bounds = self._plane_grid_boundaries
+        centers = [(bs[1:] + bs[:-1]) / 2 for bs in bounds]
+
+        # Interpolate tangential field components to cell centers
+        interp_dict = dict(zip(tan_dims, centers))
+        centered_fields = {key: val.interp(**interp_dict) for key, val in tan_fields.items()}
+        return centered_fields
+
+    @property
+    def poynting(self) -> xr.DataArray:
+        """Time-averaged Poynting vector for frequency-domain data associated to a 2D monitor."""
+
+        # Tangential fields are ordered as E1, E2, H1, H2
+        tan_fields = self._centered_tangential_fields
+        field_list = list(tan_fields.values())
+        e_x_h_star = field_list[0] * field_list[3].conj() - field_list[1] * field_list[2].conj()
+        poynting = 0.5 * np.real(e_x_h_star)
+        # cast to xr.DataArray as this is technically no longer a ScalarFieldArray
+        return xr.DataArray(poynting)
+
+    @property
+    def flux(self) -> xr.DataArray:
+        """Flux for data corresponding to a 2D monitor."""
+
+        # Compute flux by integrating Poynting vector in-plane
+        d_area = self._diff_area
+        return (self.poynting * d_area).sum(dim=d_area.dims)
+
+    def dot(
+        self, field_data: Union[FieldData, ModeSolverData], conjugate: bool = True
+    ) -> xr.DataArray:
+        """Dot product (modal overlap) with another ``field_data`` object. Both datasets have to be
+        frequency-domain data associated with a 2D monitor. Along the tangential directions,
+        the datasets have to have the same discretization. Along the normal direction, the monitor
+        position may differ and is ignored. Other coordinates (``frequency``, ``mode_index``) have
+        to be either identical or broadcastable.
+
+        Parameters
+        ----------
+        field_data : :class:`ElectromagneticFieldData`
```

Comment:
are we doing `.` before the class names? I never understood when to use that but it seems to work more often.

Reply:
This is my understanding: if you don't do the `.`, the object needs to be defined inside the current file. If you do the `.`, it needs to be imported in whatever is defined as current module [here](https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/develop/docs/source/api.rst#api-reference). So `ElectromagneticFieldData` can be without a dot (or probably even has to be since it's not in tidy3d init), while something like a `FieldMonitor` has to be with a dot. By the way that also means that there are things like `AbstractFieldDataset` which are neither in the current file nor in tidy3d init which need to be referenced in an entirely different way? But those typically won't be in our api reference either so maybe it's fine.